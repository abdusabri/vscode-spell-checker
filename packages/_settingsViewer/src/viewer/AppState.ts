import {observable, computed} from 'mobx';
import { Settings, ConfigTarget, LocalId, SettingByConfigTarget, Config, Configs, LocalList, WorkspaceFolder, TextDocument, configTargets, ConfigTargets } from '../api/settings/';
import { normalizeCode, lookupCode } from '../iso639-1';
import { compareBy, compareEach } from '../api/utils/Comparable';
import { uniqueFilter } from '../api/utils';
import { Messenger } from '../api';


type Maybe<T> = T | undefined;
type TabTargets = ConfigTarget | 'file' | 'dictionaries' | 'about';

export interface Tab {
    label: string;
    target: TabTargets;
}

const tabs: Tab[] = [
    { label: 'User', target: 'user' },
    { label: 'Workspace', target: 'workspace' },
    { label: 'Folder', target: 'folder' },
    { label: 'File', target: 'file' },
    { label: 'Dictionaries', target: 'dictionaries' },
    { label: 'About', target: 'about' },
];

export interface LanguageInfo {
    code: string;
    name: string;
    dictionaries: string[];
    enabled: boolean;
}

export interface LanguageConfig {
    languages: LanguageInfo[];
    inherited?: ConfigTarget;
}

export interface LanguageConfigs extends SettingByConfigTarget<LanguageConfig> {}

export interface State {
    activeTabName: string;
    settings: Settings;
    tabs: Tab[];
    activeTab: Tab;
    languageConfig: LanguageConfigs;
}

export interface FoundInConfig<T> {
    value: Exclude<T, undefined>,
    target: ConfigTarget
}

type InheritedFromTarget<T> = {
    value: Exclude<T, undefined>,
    target: ConfigTarget;
}


type InheritMembers<T> = {
    [K in keyof T]: InheritedFromTarget<T[K]>;
}

type InheritedConfigs = SettingByConfigTarget<Config>;

export class AppState implements State {
    @observable activeTabName = '';
    @observable settings: Settings = {
        dictionaries: [
        ],
        configs: {
            user: { inherited: {}, locals: [], languageIdsEnabled: [] },
            workspace: { inherited: { locals: 'user', languageIdsEnabled: 'user' }, locals: [], languageIdsEnabled: [] },
            folder: { inherited: { locals: 'user', languageIdsEnabled: 'user' }, locals: [], languageIdsEnabled: [] },
            file: undefined,
        }
    };
    @observable debugMode: boolean = false;

    constructor(private messageBus: Messenger) {}

    @computed get tabs() {
        const hidden = new Set<TabTargets>(configTargets.filter(target => !this.settings.configs[target]));
        if (this.workspaceFolders.length <= 1) {
            hidden.add(ConfigTargets.folder);
        }
        return tabs.filter(tab => !hidden.has(tab.target));
    }

    @computed get activeTab() {
        return this.tabs.find(t => t.label === this.activeTabName) || this.tabs[0];
    }

    @computed get languageConfig(): LanguageConfigs {
        const calcConfig = (target: ConfigTarget): LanguageConfig => {
            const config = this.settings.configs[target];
            const locals = config.locals; // todo: calc inheritance
            const inherited = config.inherited.locals;

            const infos = new Map<string, LanguageInfo>();

            const addLocalsToInfos = (locals: string[], dictionaryName: string | undefined) => {
                locals.map(normalizeCode).map(lookupCode).filter(notUndefined).forEach(lang => {
                    const { code, lang: language, country } = lang;
                    const name = country ? `${language} - ${country}` : language;
                    const found = this.isLocalEnabledEx(target, code);
                    const enabled = found && found.value || false;
                    const info: LanguageInfo = infos.get(name) || {
                        code,
                        name,
                        enabled,
                        dictionaries: [],
                    };
                    if (dictionaryName) {
                        info.dictionaries.push(dictionaryName);
                    }
                    infos.set(name, info);
                });
            }
            if (locals) {
                addLocalsToInfos(locals, undefined);
            }
            this.settings.dictionaries.forEach(dict => addLocalsToInfos(dict.locals, dict.name));

            return {
                languages: [...infos.values()].sort(compareEach(
                    compareBy(info => !info.dictionaries.length),
                    compareBy('name'),
                )),
                inherited
            };
        };

        return {
            user: calcConfig('user'),
            workspace: calcConfig('workspace'),
            folder: calcConfig('folder'),
        }
    }

    @computed get activeTabIndex(): number {
        const index = this.tabs.findIndex(t => t.label === this.activeTabName);
        return index > 0 ? index : 0;
    }

    @computed get workspaceFolderNames(): string[] {
        const workspace = this.settings.workspace;
        const folders = workspace && workspace.workspaceFolders || [];
        return folders.map(f => f.name);
    }

    @computed get activeWorkspaceFolder(): string | undefined {
        const folder = this.findMatchingFolder(this.settings.activeFolderUri);
        return folder && folder.name;
    }

    @computed get workspaceFolders(): WorkspaceFolder[] {
        const workspace = this.settings.workspace;
        return workspace && workspace.workspaceFolders || [];
    }

    @computed get activeFileUri(): string | undefined {
        return this.settings.activeFileUri;
    }

    @computed get documents(): TextDocument[] {
        const workspace = this.settings.workspace;
        return workspace && workspace.textDocuments || [];
    }

    @computed get documentSelection(): { label: string, value: string }[] {
        return this.documents.map(doc => ({ label: doc.fileName, value: doc.uri }));
    }

    private findMatchingFolder(uri: string | undefined): WorkspaceFolder | undefined {
        return this.workspaceFolders.filter(f => f.uri === uri)[0];
    }

    private findMatchingFolderByName(name: string | undefined): WorkspaceFolder | undefined {
        return this.workspaceFolders.filter(f => f.name === name)[0];
    }

    actionSetLocal(target: ConfigTarget, local: LocalId, enable: boolean) {
        const localsCurrent = this.settings.configs[target].locals;
        const locals = (enable ? [...localsCurrent, local] : localsCurrent.filter(v => v !== local)).filter(uniqueFilter());
        this.settings.configs[target].locals = locals;
        this.messageBus.postMessage({ command: 'EnableLocalMessage', value: { target, local, enable } });
    }

    actionSetDebugMode(isEnabled: boolean) {
        this.debugMode = isEnabled;
    }

    private isLocalEnabledEx(field: ConfigTarget, code: LocalId):InheritedFromTarget<boolean> {
        const config = this.settings.configs[field];
        const target = config.inherited.locals || field;
        const locals = config.locals;
        return  {
            value: locals.map(normalizeCode).includes(code),
            target
        };
    }


    actionActivateTabIndex(index: number) {
        const tab = this.tabs[index];
        if (tab) {
            this.actionActivateTab(tab.label);
        }
    }

    actionActivateTab(tabName: string) {
        this.activeTabName = tabName;
        this.messageBus.postMessage({ command: 'SelectTabMessage', value: this.activeTabName });
    }

    actionSelectFolder(folderName: string) {
        const folder = this.findMatchingFolderByName(folderName);
        const folderUri = folder && folder.uri;
        this.settings.activeFolderUri = folderUri;
        if (folderUri !== undefined) {
            this.messageBus.postMessage({ command: 'SelectFolderMessage', value: folderUri })
        }
    }

    actionSelectDocument(documentUri: string) {
        this.messageBus.postMessage({ command: 'SelectFileMessage', value: documentUri });
    }

    actionEnableLanguageId(languageId: string, enable: boolean, target?: ConfigTarget) {
        const fileConfig = this.settings.configs.file;
        if (fileConfig && fileConfig.languageId === languageId) {
            fileConfig.languageEnabled = enable;
        }
        this.messageBus.postMessage({ command: 'EnableLanguageIdMessage', value: { languageId, enable, target }});
    }
}

function notUndefined<T>(a : T): a is Exclude<T, undefined> {
    return a !== undefined;
}


