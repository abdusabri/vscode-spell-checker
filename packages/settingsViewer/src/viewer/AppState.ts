import {observable, computed} from 'mobx';
import { Settings } from '../api/settings/';

export const cats = [
    { title: 'Coding Cat', image: 'https://media.giphy.com/media/JIX9t2j0ZTN9S/giphy.gif', icon: 'home'},
    { title: 'Compiling Cat', image: 'https://media.giphy.com/media/mlvseq9yvZhba/giphy.gif', icon: 'face'},
    { title: 'Testing Cat', image: 'https://media.giphy.com/media/3oriO0OEd9QIDdllqo/giphy.gif', icon: 'favorite'},
];

export const tabLabels = ['Language'].concat(cats.map(cat => cat.title));

export class AppState {
    @observable activeTabIndex = 0;
    @observable timer = 0;
    @observable counter = 0;
    @observable settings: Settings = {
        locals: [
            {
                code: 'en',
                name: 'English',
                dictionaries: ['en', 'en-us'],
                enabled: true,
                isInUserSettings: true,
            },
            {
                code: 'es',
                name: 'Spanish',
                dictionaries: ['es', 'es-ES'],
                enabled: true,
                isInUserSettings: true,
            },
        ],
    };

    @computed get activeTab() {
        return tabLabels[this.activeTabIndex];
    }

    constructor() {
        setInterval(() => {
            this.timer += 1;
        }, 1000);
    }

    resetTimer() {
        this.timer = 0;
    }
}
