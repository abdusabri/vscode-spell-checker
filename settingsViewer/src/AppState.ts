import {observable, computed} from 'mobx';
import { ConfigurationForDocument } from './settings/';

export const cats = [
    { title: 'Coding Cat', image: 'https://media.giphy.com/media/JIX9t2j0ZTN9S/giphy.gif', icon: 'home'},
    { title: 'Compiling Cat', image: 'https://media.giphy.com/media/mlvseq9yvZhba/giphy.gif', icon: 'face'},
    { title: 'Testing Cat', image: 'https://media.giphy.com/media/3oriO0OEd9QIDdllqo/giphy.gif', icon: 'favorite'},
];

export const tabLabels = ['Language'].concat(cats.map(cat => cat.title));

export class AppState {
    @observable activeTabIndex = 0;
    @observable timer = 0;
    @observable counter = 0;
    @observable config: ConfigurationForDocument = {
        languageEnabled: true,
        fileEnabled: true,
        settings: {},
        docSettings: {}
    };
    @computed get activeTab() {
        return tabLabels[this.activeTabIndex];
    }

    constructor() {
        setInterval(() => {
            this.timer += 1;
        }, 1000);
    }

    resetTimer() {
        this.timer = 0;
    }
}
