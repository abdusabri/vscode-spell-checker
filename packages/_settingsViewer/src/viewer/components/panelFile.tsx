import * as React from 'react';
import {observer} from 'mobx-react';
import { AppState } from '../AppState';
import Select from '@material/react-select';
import { toJS } from 'mobx';
import { SectionDictionaries } from './sectionDictionaries';
import { Grid, Row, Cell } from '@material/react-layout-grid';
import List, { ListItem, ListItemGraphic, ListItemText, ListItemMeta } from '@material/react-list';
import { Checkbox } from '@material/react-checkbox';
import MaterialIcon from '@material/react-material-icon';

function initRipple(){}

@observer
export class PanelFile extends React.Component<{appState: AppState}, {}> {
    render() {
        const appState = this.props.appState;
        const settings = appState.settings;
        const config = settings.configs.file;
        const languageId = config && config.languageId || 'unknown';
        const languageEnabled = config && config.languageEnabled;
        const fileEnabled = config && config.fileEnabled;
        const dictionaries = config && config.dictionaries || [];
        const select = (elem: HTMLSelectElement) => elem && this.props.appState.actionSelectDocument(elem.value);
        return (
            <div>
                <h2>
                    File
                </h2>
                <Select className='select_folder'
                        label='File'
                        options={appState.documentSelection}
                        onChange={(evt) => select(evt.target as HTMLSelectElement)}
                        value={appState.activeFileUri}
                    />
                <h2>Settings</h2>
                {/*
                <Grid>
                    <Row>
                        <Cell columns={8}>
                        </Cell>
                    </Row>
                </Grid>
                */}
                        <List>
                            <ListItem role='checkbox'>
                                <ListItemGraphic graphic={<MaterialIcon icon='description'/>} />
                                <ListItemText primaryText={`File enabled`} />
                                <ListItemMeta meta={
                                    <Checkbox
                                        disabled={true}
                                        checked={fileEnabled}
                                        initRipple={initRipple} />
                                }/>
                            </ListItem>
                            <ListItem role='checkbox' shouldToggleCheckbox={true} onClick={evt => this.enableLanguageId(!languageEnabled)}>
                                <ListItemGraphic graphic={<MaterialIcon icon='code'/>} />
                                <ListItemText primaryText={<label htmlFor='checkbox-language-enabled'>{`Programming Language: ${languageId}`}</label>} />
                                <ListItemMeta meta={
                                    <Checkbox
                                        nativeControlId='checkbox-language-enabled'
                                        checked={languageEnabled}
                                        onChange={(evt) => this.enableLanguageId(evt.target.checked)}
                                        initRipple={initRipple} />
                                }/>
                            </ListItem>
                        </List>

                <SectionDictionaries dictionaries={dictionaries} sectionTitle='Active Dictionaries'></SectionDictionaries>
                {appState.debugMode ?
                    <div>
                        <pre>{JSON.stringify(toJS(config), null, 2)}</pre>
                    </div>
                : ''}
            </div>
        );
    }

    enableLanguageId(enable: boolean) {
        const config = this.props.appState.settings.configs.file;
        const languageId = config && config.languageId;
        if (languageId) {
            this.props.appState.actionEnableLanguageId(languageId, enable);
        }
    }
}
