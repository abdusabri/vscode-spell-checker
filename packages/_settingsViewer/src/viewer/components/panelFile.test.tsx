
import * as React from 'react';
import { AppState } from '../AppState';
import { create } from 'react-test-renderer';
import { sampleAppState } from '../../test/fixtures/AppState';
import { PanelFile } from './panelFile';

describe('File Panel Verification', () => {
    it('tests the snapshot', () => {
        const appState = getSampleAppState();
        const panelRenderer = create(<PanelFile appState={appState}></PanelFile>).toJSON()!;
        expect(panelRenderer).toMatchSnapshot();
    });

    function getSampleAppState(): AppState {
        return sampleAppState();
    }
});
