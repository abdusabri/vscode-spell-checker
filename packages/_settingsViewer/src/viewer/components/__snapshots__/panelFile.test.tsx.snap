// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File Panel Verification tests the snapshot 1`] = `
<div>
  <h2>
    File
  </h2>
  <div
    className="mdc-select select_folder"
  >
    <select
      className="mdc-select__native-control"
      disabled={false}
      label="File"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onTouchStart={[Function]}
      options={
        Array [
          Object {
            "label": "ripgrep/.vscode/settings.json",
            "value": "file:///Users/cspell/projects/clones/ripgrep/.vscode/settings.json",
          },
          Object {
            "label": "vscode-cspell-dict-extensions/extensions/dutch/src/extension.ts",
            "value": "file:///Users/cspell/projects/vscode-cspell-dict-extensions/extensions/dutch/src/extension.ts",
          },
          Object {
            "label": "ripgrep/src/decoder.rs",
            "value": "file:///Users/cspell/projects/clones/ripgrep/src/decoder.rs",
          },
        ]
      }
      value="file:///Users/cspell/projects/vscode-cspell-dict-extensions/extensions/dutch/src/extension.ts"
    >
      <option
        value="file:///Users/cspell/projects/clones/ripgrep/.vscode/settings.json"
      >
        ripgrep/.vscode/settings.json
      </option>
      <option
        value="file:///Users/cspell/projects/vscode-cspell-dict-extensions/extensions/dutch/src/extension.ts"
      >
        vscode-cspell-dict-extensions/extensions/dutch/src/extension.ts
      </option>
      <option
        value="file:///Users/cspell/projects/clones/ripgrep/src/decoder.rs"
      >
        ripgrep/src/decoder.rs
      </option>
    </select>
    <label
      className="mdc-floating-label mdc-floating-label--float-above"
      onAnimationEnd={[Function]}
    >
      File
    </label>
    <div
      className="mdc-line-ripple"
      onTransitionEnd={[Function]}
      style={Object {}}
    />
  </div>
  <h2>
    Settings
  </h2>
  <ul
    aria-orientation="vertical"
    className="mdc-list"
  >
    <li
      className="mdc-list-item"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      role="checkbox"
      tabIndex={0}
    >
      <i
        className="material-icons mdc-list-item__graphic"
        tabIndex={-1}
      >
        description
      </i>
      <span
        className="mdc-list-item__text"
        tabIndex={-1}
      >
        File enabled
      </span>
      <div
        className="mdc-checkbox mdc-checkbox--upgraded mdc-checkbox--disabled mdc-list-item__meta"
        onAnimationEnd={[Function]}
        tabIndex={-1}
      >
        <input
          aria-checked={false}
          checked={true}
          className="mdc-checkbox__native-control"
          disabled={true}
          onChange={[Function]}
          type="checkbox"
        />
        <div
          className="mdc-checkbox__background"
        >
          <svg
            className="mdc-checkbox__checkmark"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              className="mdc-checkbox__checkmark-path"
              d="M1.73,12.91 8.1,19.28 22.79,4.59"
              fill="none"
            />
          </svg>
          <div
            className="mdc-checkbox__mixedmark"
          />
        </div>
      </div>
    </li>
    <li
      className="mdc-list-item"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      role="checkbox"
      tabIndex={-1}
    >
      <i
        className="material-icons mdc-list-item__graphic"
        tabIndex={-1}
      >
        code
      </i>
      <label
        className="mdc-list-item__text"
        htmlFor="checkbox-language-enabled"
      >
        Programming Language: typescript
      </label>
      <div
        className="mdc-checkbox mdc-checkbox--upgraded mdc-list-item__meta"
        onAnimationEnd={[Function]}
        tabIndex={-1}
      >
        <input
          aria-checked={false}
          checked={true}
          className="mdc-checkbox__native-control"
          disabled={false}
          id="checkbox-language-enabled"
          onChange={[Function]}
          type="checkbox"
        />
        <div
          className="mdc-checkbox__background"
        >
          <svg
            className="mdc-checkbox__checkmark"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              className="mdc-checkbox__checkmark-path"
              d="M1.73,12.91 8.1,19.28 22.79,4.59"
              fill="none"
            />
          </svg>
          <div
            className="mdc-checkbox__mixedmark"
          />
        </div>
      </div>
    </li>
  </ul>
  <div>
    <h2>
      Active Dictionaries
    </h2>
    <ul
      aria-orientation="vertical"
      className="mdc-list dictionary_list mdc-list--two-line"
    >
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          code
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            node
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            List of NodeJS terms.
          </span>
        </span>
        <span
          className="mdc-list-item__meta"
          tabIndex={-1}
        >
          
        </span>
      </li>
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        tabIndex={-1}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          code
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            npm
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            List of Top 500 NPM packages.
          </span>
        </span>
        <span
          className="mdc-list-item__meta"
          tabIndex={-1}
        >
          
        </span>
      </li>
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        tabIndex={-1}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          code
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            typescript
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            JavaScript and Typescript terms.
          </span>
        </span>
        <span
          className="mdc-list-item__meta"
          tabIndex={-1}
        >
          
        </span>
      </li>
    </ul>
  </div>
  
</div>
`;
